core/src/main/java/com/linecorp/armeria/common/CompositeStringMultimap.java

```diff
@@ -0,0 +1,1111 @@
+            }
+        }
+        return false;
+    }
+
+    @Override
+    public boolean containsDouble(IN_NAME name, double value) {
+        requireNonNull(name, "name");
+        if (removed.contains(name)) {
+            return appender != null && appender.containsDouble(name, value);
+        }
+
+        for (StringMultimapGetters<IN_NAME, NAME> delegate : delegates()) {
+            if (delegate.containsDouble(name, value)) {
+                return true;
+            }
+        }
+        return false;
+    }
+
+    @Override
+    public boolean containsTimeMillis(IN_NAME name, long value) {
+        requireNonNull(name, "name");
+        if (removed.contains(name)) {
+            return appender != null && appender.containsTimeMillis(name, value);
+        }
+
+        for (StringMultimapGetters<IN_NAME, NAME> delegate : delegates()) {
+            if (delegate.containsTimeMillis(name, value)) {
+                return true;
+            }
+        }
+        return false;
+    }
+
+    @Override
+    public int size() {
+        int size = 0;
+        for (StringMultimapGetters<IN_NAME, NAME> delegate : delegates()) {
+            size += delegate.size();
+        }
+
+        if (removed.isEmpty()) {
+            return size;
+        }
+
+        for (IN_NAME name : removed) {
+            final Integer cachedRemovedSize = removedSizeCache.get(name);
+            if (cachedRemovedSize != null) {
+                size -= cachedRemovedSize;
+                continue;
+            }
+
+            int removedSize = 0;
+            for (StringMultimapGetters<IN_NAME, NAME> parent : parents) {
+                if (parent.contains(name)) {
+                    removedSize += parent.getAll(name).size();
+                }
+            }
+            size -= removedSize;
+            removedSizeCache.put(name, removedSize);
+        }
+
+        return Math.max(size, 0);
+    }
+
+    @Override
+    public boolean isEmpty() {
+        if (appender != null && !appender.isEmpty()) {
+            return false;
+        }
+
+        for (StringMultimapGetters<IN_NAME, NAME> parent : parents) {
+            for (NAME name : parent.names()) {
+                if (!removed.contains(name.toString())) {
+                    return false;
+                }
+            }
+        }
+        return true;
+    }
+
+    @Override
+    public final Set<NAME> names() {
+        final ImmutableSet.Builder<NAME> builder = ImmutableSet.builder();
+        for (StringMultimapGetters<IN_NAME, NAME> parent : parents) {
+            for (NAME name : parent.names()) {
+                if (!removed.contains(name.toString())) {
+                    builder.add(name);
+                }
+            }
+        }
+
+        if (appender != null && !appender.isEmpty()) {
+            builder.addAll(appender.names());
+        }
+        return builder.build();
+    }
+
+    @Override
+    public Iterator<Entry<NAME, String>> iterator() {
```

Comment:
Question) To match the current merge logic, should this loop deduplicate such that only one key is returned?

i.e.
```
final CompositeStringMultimap<CharSequence, AsciiString> headers =
        new CompositeHttpHeadersBase(HttpHeaders.of("k1", "v1"),
                                     HttpHeaders.of("k1", "v2"));
for (Entry<AsciiString, String> entry: headers) {
    System.out.println(entry); // ("k1", "v1") only should be printed
}
```

ditto for the other iterating methods (`forEach`, `forEachValue`, etc..)

Reply:
https://github.com/line/armeria/blob/d328e270cd9955929632bb55a12ed694c92ee574/core/src/main/java/com/linecorp/armeria/internal/common/HttpHeadersUtil.java#L46-L48

https://github.com/line/armeria/blob/d328e270cd9955929632bb55a12ed694c92ee574/core/src/main/java/com/linecorp/armeria/internal/common/HttpHeadersUtil.java#L64-L66

Oh you're right! based on current merge logic, we don't add duplicated key.

I think `CompositeHeaders` can support `merge (remove duplicated key)` and `just composite` options.
(or maybe we can explicitly introduce `CompositeHeader` and `MergedHeader`)

Anyway, I'll more think about this design and share to you soon~! 🙇

Reply:
```java
// 1) Merge headers same with HttpHeadersUtil.merge(additionalHeaders, headers, defaultHeaders)
CompositeHeader additionals = new CompositeHeader(add1, add2, ..);
for (Name name : additional.names() {
  if (isPsedoHeader(name) ||
      ADDITIONAL_REQUEST_HEADER_DISALLOWED_LIST.contains(name)) {
    additional.remove(name) // Only mark as removed, without deep copy
  }
}

..

// Can composite default and internal header like this without deep copy
HttpHeaders internals = HttpHeaders.of(HttpHeaderNames.SERVER, ArmeriaHttpUtil.SERVER_HEADER);
CompositeHeader defaults = new CompositeHeader(default1, internals);

// 2) Also, can composite headers
CompositeHeader headers = new CompositeHeader(header1, header2, ..);
```
✅ To respect current `HttpHeadersUtil.merge(additionals, headers, defaults)` logic,
I make `CompositeHeader` can merge `additional, header, default` too with same logic, without deep copy~!